sklearn/tree/_criterion.pyx

```diff
@@ -203,11 +205,17 @@ cdef class Criterion:
         self.children_impurity(&impurity_left, &impurity_right)
 
         return ((self.weighted_n_node_samples / self.weighted_n_samples) *
-                (impurity - (self.weighted_n_right / 
+                (impurity - (self.weighted_n_right /
```

Comment:
also here you could overwrite impurity_improvement function in a subclass and add init_data as its constructor instead of modifying the main one